zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1221,20 +1237,109 @@ private void reconnect() {
     }
   }
 
+  private class SyncCallbackHandler extends ZkAsyncCallbacks.DefaultCallback implements AsyncCallback.VoidCallback {
+    private String _sessionId;
+
+    SyncCallbackHandler(String sessionId) {
+      _sessionId = sessionId;
+    }
+
+    @Override
+    public void processResult(int rc, String path, Object ctx) {
+      LOG.info("sycnOnNewSession with sessionID {} async return code: {}", _sessionId, rc);
+      callback(rc, path, ctx);
+    }
+
+    @Override
+    public void handle() {
+      // Make compiler happy, not used.
+    }
+
+    @Override
+    protected boolean needRetry(int rc) {
+      try {
+        switch (KeeperException.Code.get(rc)) {
+          /** Connection to the server has been lost */
+          case CONNECTIONLOSS:
+            return true;
+          default:
+            return false;
+        }
+      } catch (ClassCastException | NullPointerException ex) {
+        LOG.error("Session {} failed to handle unknown return code {}. Skip retrying. ex {}",
+            _sessionId, rc, ex);
+        return false;
+      }
+    }
+  }
+
+  private void doAsyncSync(final ZooKeeper zk, final String path, final long startT,
+      final SyncCallbackHandler cb) {
+    zk.sync(path, cb,
+        new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, false) {
+          @Override
+          protected void doRetry() throws Exception {
+            doAsyncSync(zk, path, startT, cb);
+          }
+        });
+  }
+
+  private boolean retrySync(String sessionId) throws ZkInterruptedException {
+    if (!_syncOnNewSession) {
+      return true;
+    }
+
+    SyncCallbackHandler callbackHandler = new SyncCallbackHandler(sessionId);
+
+    final ZkConnection zkConnection = (ZkConnection) getConnection();
```

Comment:
I think you can just refer the ZkClient object itself in the doAsyncSync call to simplify these codes. Or is there any concern of reading the Zookeeper inside doAsyncSync where it is really used?

Reply:
I guess you concern if the Zookeeper object is changed if new session established. But I think we can still keep retrying the sync call, although it is in theory not possible because we stop retrying on session expiring.
Or is there anything else?

Reply:
You got the point. One zookeeper object one session, This is the invariant.

For this fixed session/zookeeper object, if sync() failed, let it be. The next time session reconnected, a new fire all event would be issued and a new retrySync() would be used with the new session.

Reply:
You mean it is ok to NOT pass the Zookeeper? The interface is awkward in the current way. You shall be able to get the Zookeeper inside the doSync call.

Reply:
I mean we should pass in ZooKeeper object for doAsyncSync(). 

This is different from other doAsyc{verb}.

Others have auto retry after session changes (thus ZooKeeper object change). But this one we need to use the same ZooKeeper object.

This is the reason why this doAsycSync() has one more parameter as Zookeeper zk. It looks different from others.

Reply:
This is my confusing part. Your retry will stop if session expired. Then even you get the zk from ZkClient every time, you will get the same one, no?

Reply:
This is how I look at it. 

Let us say the code is like this 

```
void doAsyncSync final String path, final long startT,
      final ZkAsyncCallbacks.SyncCallbackHandler cb) {

     final ZooKeeper zk = (ZkConnection) getConnection()).getZookeeper();
     zk.sync(path, cb,
        new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, false) {
          @Override
          protected void doRetry() throws Exception {
            doAsyncSync(zk, path, System.currentTimeMillis(), cb);
          }
        });
  }

```

Then, there is chance that when connection_loss, we have retry request from the asyncThread of ZkClient. The retry request is finally doAsyncSync(). Then, it would retrieve the ZooKeeper object from connection again. Note, by this time, the session can be changed (thus ZooKeeper object with a new session ID) is used. 

This is not what we want. At least this sync() call is wasted for the reason the new session would invoke another fireAllEvent().

Reply:
I got what you said, and I agree with the idea. But this is hard to achieve, even with your current work, the zookeeper object could be changed just before this selected line.
To make it solid,
1. We need to check the session Id after you get the zk object. If session does not match then skip the operation.
2. You might want to try-catch the line 1260 since if the zk already becomes invalid at the first call, it might throw an exception and we can just return false.

In addition, I think it is a good point to comment on the code so new devs won't "simplify" this part. And let's have a test to cover this case to ensure the callback wait won't be stuck. In detail, the zk object is changed during the retry.

Reply:
> 1. We need to check the session Id after you get the zk object. If session does not match then skip the operation.

The first ZooKeeper object extracted from connection (line 1259) is up in the stack in processStateChange() and then passed into doAsync(). Note, procesStateChange() is inside the lock of getEventLock(). I think under the lock, there is not possibility of changing another ZooKeeper object of different session, right? So this is not a concern in this case.

> 2. You might want to try-catch the line 1260 since if the zk already becomes invalid at the first call, it might throw an exception and we can just return false.

See the reply above. Am I missing something here? Can you give an example?

Reply:
processStateChange() and fireNewSessionEvents() is in the lock for sure.
However, your retrySync() is in the event thread. When it is running, it is not in any ZkEventLock protection. If you have 2 consequence new session establishments, then you might have 2 events in the queue if the processing is stuck. And when you process the first new session event, you are using the latest zk object (the wrong one).
That is exactly why we check the session Id in the ZkHelixManager.handleNewSession().

Feel free to refer to the related new session handling test cases for the similar logic.

Reply:
This is a good point. My intention is actually to get the Zookeeper object from Zookeeper client event thread and pass it to retrySync().

```
    if (_syncOnNewSession) {

      final ZkConnection zkConnection = (ZkConnection) getConnection();
      final ZooKeeper zk = zkConnection.getZookeeper();
      _eventThread.send(new ZkEventThread.ZkEvent("Sync call before new session event of session " + sessionId,
          sessionId) {
        @Override
        public void run() throws Exception {
          if (retrySync(sessionId, zk) == false) {
            LOG.warn("sync on session {} failed", sessionId);
          }
        }
      });
    }
```

Reply:
That works. But you already have sessionId. You can check and decide whether to skip the run or not. Why bother passing the redundant information?

Reply:
Check sessionId is after the fact. It can incur unnecessary sync() to server. For the new session, there may be two sync() in a row. 

In this case we are asserting that for each new session, we have exactly one sync() going to server.

Reply:
Added test and comments

Reply:
Sorry that I cannot follow the comment that "Check sessionId is after the fact...".
Could you please give me examples that how this invokes multiple sync() to the server? And how your current design prevent it.
TO be clear what I meant by adding the session check is that check session Id in retrySync() before doAsyncSync.

`
final ZkConnection zkConnection = (ZkConnection) getConnection();
final ZooKeeper zk = zkConnection.getZookeeper();
if (zk.getSessionId().equals(sessionId)) {
  doAsyncSync(zk, _syncPath, startT, callbackHandler);
} else {
  return false; // or throw exception
}
`

Reply:
Just saw this comment. 
The gist is that the checking of sessionId and doAsyncSync() is not atomic. After `if (zk.getSessionId().equals(sessionId))` succeed, the Zookeeper object may be changed due to a new session established. This would make the current doAsyncSync() work for the new session. Since the new session would trigger another fireNewSession(), which would put another  retrySync(). Then we will have another sync() to the Zookeeper for the new session. So in sum, the new session can be two sync() in a row.

Reply:
I will argue that it won't be an issue even it is not atomic. Since you get the zk object first. So if the zk object is changed, the later check will not be applied to the newer zk object.